pip_faster.py

```diff
@@ -41,209 +47,145 @@
-
-        buildset = [
-            req for req in reqset
-            # FASTER: don't wheel things that have no source available
-            if wheelable(req)
-        ]
-
-        if not buildset:
-            return buildset
-
-        # build the wheels
-        logger.notify(
-            'Building wheels for collected packages: %s' %
-            ','.join([req.name for req in buildset])
-        )
-        for req in buildset:
-            self._build_one(req)
+def mkdirp(pth):
+    try:
+        os.makedirs(pth)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            raise
 
-        return buildset
 
+def cache_installed_wheels(index_url, installed_packages):
+    """After installation, pip tells us what it installed and from where.
 
-def pipfaster_packagefinder():
-    """Provide a short-circuited search when the requirement is pinned and appears on disk.
+    We build a structure that looks like
 
-    Suggested upstream at: https://github.com/pypa/pip/pull/2114
+    .cache/pip-faster/wheelhouse/$index_url/$package/$version/$wheel ->
+        (the path pip tells us)
     """
-    # A poor man's dependency injection: monkeypatch :(
-    # we need this exact import -- pip clobbers `commands` in their __init__
-    from pip.commands import install
-    return patched(vars(install), {'PackageFinder': FasterPackageFinder})
-
-
-@contextmanager
-def pipfaster_install():
-    # pip<6 needs this exact import -- they clobber `commands` in __init__
-    from pip.commands import install
-    with patched(vars(install), {'RequirementSet': FasterRequirementSet}):
-        @property
-        def delete_marker_filename(self):
-            """
-            pip.req.InstallRequirement.delete_marker_filename contains an unhelpful assertion of `self.source_dir`
-            Instead, we return a filename that will never exist.
-            """
-            if not self.source_dir:
-                return ''  # The empty string never exists: http://man7.org/linux/man-pages/man2/stat.2.html#ERRORS
-            else:
-                return orig.__get__(self)  # pylint: disable=no-member
-
-        from pip.req import InstallRequirement
-        orig = InstallRequirement.delete_marker_filename
-        InstallRequirement.delete_marker_filename = delete_marker_filename
-        yield
-        InstallRequirement.delete_marker_filename = orig
+    for installed_package in installed_packages:
+        if not _can_be_cached(installed_package):
+            continue
+        dest = installed_package.link.path
+        filename = os.path.basename(dest)
+        wheel = Wheel(filename)
+        cache = os.path.join(
+            CacheOpts().wheelhouse,
+            index_url,
+            wheel.name.lower(),
+            filename,
+        )
+        cache_dir = os.path.dirname(cache)
+        mkdirp(cache_dir)
+        rel = os.path.relpath(dest, cache_dir)
+        subprocess.check_call(('ln', '-sf', rel, cache))
```

Comment:
could this have a race condition? worth doing the 'make symlink, mv into correct location' thing?